src/librustc_codegen_llvm/intrinsic.rs

```diff
@@ -141,26 +141,17 @@ impl IntrinsicCallMethods<'tcx> for Builder<'a, 'll, 'tcx> {
                 self.call(llfn, &[], None)
             }
             "count_code_region" => {
-                if let ty::InstanceDef::Item(fn_def_id) = caller_instance.def {
-                    let caller_fn_path = tcx.def_path_str(fn_def_id);
-                    debug!(
-                        "count_code_region to llvm.instrprof.increment(fn_name={})",
-                        caller_fn_path
-                    );
-
-                    // FIXME(richkadel): (1) Replace raw function name with mangled function name;
-                    // (2) Replace hardcoded `1234` in `hash` with a computed hash (as discussed in)
-                    // the MCP (compiler-team/issues/278); and replace the hardcoded `1` for
-                    // `num_counters` with the actual number of counters per function (when the
-                    // changes are made to inject more than one counter per function).
-                    let (fn_name, _len_val) = self.const_str(Symbol::intern(&caller_fn_path));
-                    let index = args[0].immediate();
-                    let hash = self.const_u64(1234);
-                    let num_counters = self.const_u32(1);
-                    self.instrprof_increment(fn_name, hash, num_counters, index)
-                } else {
-                    bug!("intrinsic count_code_region: no src.instance");
-                }
+                let coverage_data = fx.mir.coverage_data.as_ref().unwrap();
```

Comment:
I think instead of directly accessing the MIR body here, we should add a new query to `TyCtxt` that can return the coverage data for given `DefId`. Then you won't need to take the `FunctionCx` here or make `FunctionCx.mir` or `FunctionCx.instance` public or add the `coverage_data` field to `Body`. 

That integrates better with our incremental system and wouldn't require these other changes.

Reply:
Great idea. Thanks!

Reply:
@wesleywiser I'm a little unsure what you are suggesting for the query, particularly in the context of _not_ storing the `coverage_data` in the `mir::Body`.

I'm trying to familiarize myself with the queries and provider implementations, but I don't know how to set it up yet.

If I implement a `query coverage_data(DefId) -> mir::CoverageData`, but create the `CoverageData` during the `InstrumentCoverage` pass, where do I save the `CoverageData` so that the query implementation will return it?

It seems like it might be possible to store it directly in something like the rustc_metadata::rmeta tables, such as:

```rust
    coverage_data: Table<DefIndex, Lazy<mir::CoverageData>>,
```

or maybe add a new Map/cache directly to `GlobalCtxt`?

But there are probably other options, and in all of these cases, I can't pinpoint any examples in MIR-time processing where data like this is computed, stored, and then retrieved from a `query` downstream, without invoking the query to compute the value.

Any suggestions or code references I can model this after?

Thanks!
Rich

Reply:
So, what I was envisioning was a two-step approach where

- the `InstrumentCoverage ` MIR pass would only inject counters and nothing else

- the `coverage_data` query would get the MIR, count how many counters are in the MIR and hash the HIR body

The result of the `coverage_data` query would then be automatically persisted for you in the compiler session context. If you ever need that data cross-crate in the future, this would also set you up to be able to do that quite easily.

Reply:
FYI, I was uploading a new commit just as you sent your reply.

Got it. I understand how that would work in the current context, but I wonder if this will have the best performance in the future.

I'm thinking of 2 or 3 possible upcoming changes that might still make it less efficient to split computing the state data (in the query) from instrumenting the MIR (in the pass). For example:

1. Just re-counting the injected calls requires walking the MIR again.
2. The implementation of the hash may change (Note that clang's approach does some coverage-specific analysis, so the hash is less sensitive to changes that don't affect flow).

With that in mind, is it reasonable to stash the coverage data in the `Body` struct, or do you feel strongly that this state data should be computed separately from the coverage analysis?

Reply:
(Note, the new commit addresses part of your recommendation; it does implement the `query`, so I don't need to change the `codegen_intrinsic_call()` function signature, and can now retrieve the CoverageData from `tcx.coverage_data(def_id)`, which is an improvement. But the query gets the CoverageData from the `Body`.)

Reply:
@wesleywiser - @tmandry  and I discussed this. We're good with either the current approach to implementing the query, or counting the intrinsic calls by traversing the MIR in the query as per your last suggestion. Once we know your preference, I believe Tyler is ready to approve.

Reply:
Thanks for being flexible about this @richkadel! I talked a bit about this with @oli-obk and we think for now not storing the `CoverageData` in the `Body` and instead re-walking the MIR is the better architecture. If we discover that causes performance issues, we'd certainly be open to a different approach to improve performance but at this time, we don't think there will be much of a performance hit and it integrated better with the rest of the compiler.

Reply:
Will do. Thanks!

Reply:
Done. PTAL. Thanks!

Reply:
This looks great to me. Thank you!